cmd/pinniped-proxy/src/logging.rs

```diff
@@ -0,0 +1,108 @@
+use std::cmp::PartialEq;
+use std::fmt;
+
+use hyper::{Body, Request, Response};
+
+#[derive(Clone, Debug, Default, PartialEq)]
+/// LogData encapsulates all the data we want to collect during a proxied request.
+///
+/// This includes data from the initial request as well as the returned response.
+pub struct LogData {
+    target_uri: hyper::Uri,
+    method: hyper::Method,
+    // status_code is optional since we don't know it's value until the upstream
+    // response is returned.
+    status_code: Option<hyper::StatusCode>,
+}
+
+/// We implement the Display trait for our LogData struct so that it
+/// knows how to display itself as a string when requested.
+impl fmt::Display for LogData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // If we have a status, use the string version of that,
+        // otherwise we'll just display a hyphen.
+        let status: String = match self.status_code {
+            Some(sc) => sc.to_string(),
+            None => "-".into(),
+        };
+        write!(f, "{} {} {}", self.method, self.target_uri, status)
+    }
+}
+
+/// request_log_data returns a new LogData with the relevant data from the request. 
+pub fn request_log_data(req: &Request<Body>) -> LogData {
+    LogData {
+        target_uri: req.uri().clone(),
+        method: req.method().clone(),
+        status_code: None,
+    }
+}
+
+/// response_log_data returns a new LogData including the relevant data from the response.
+pub fn response_log_data(resp: &Response<Body>, log_data: LogData) -> LogData {
+    let mut updated_log_data = log_data.clone();
+    updated_log_data.status_code = Some(resp.status());
+    updated_log_data
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Tests in rust modules are included as a sub module within the same file. With this change, running `cargo test` results in:

```
$ cargo test
   Compiling pinniped-proxy v0.1.0 (/home/michael/dev/kubeapps/cmd/pinniped-proxy)
    Finished test [unoptimized + debuginfo] target(s) in 2.17s
     Running target/debug/deps/pinniped_proxy-6b227f04e120477c

running 4 tests
test logging::tests::display_including_status ... ok
test logging::tests::display_without_status ... ok
test logging::tests::populate_from_request ... ok
test logging::tests::populate_from_response ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

Reply:
Is this approach typical in Rust? I mean, why not having the tests in a separate file?

Reply:
One of the benefits of having it as a sub module is that you can `use super::*` to have access to all the private types etc. So it appears that unit tests always go in the same file - https://doc.rust-lang.org/book/ch11-03-test-organization.html#the-tests-module-and-cfgtest . Integration tests go in a `tests` directory: https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests